[CSDiagnostics] Use special error message when call didn't match exactly

xedin opened the pull request:
For multiple solutions with fixes for the same call, replace
`ambiguous reference` diagnostic with the one that explicitly
mentions that there are no exact matches, and provide partially
matched candidates as notes.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[CSDiagnostics] Use special error message when call didn't match exactly

For multiple solutions with fixes for the same call, replace
`ambiguous reference` diagnostic with the one that explicitly
mentions that there are no exact matches, and provide partially
matched candidates as notes.

xedin commented:
Message is the best I could come up with, input is appreciated @jrose-apple.

jrose-apple commented:
Which kind of ambiguity is this? Is it possible for the labels to mismatch? If so, it's not really an "overload". (We have a longstanding bug where one of our diagnostics prints the full name of the first match even when several matches have different argument labels, but I don't remember if it's *this* diagnostic.)

xedin commented:
@jrose-apple Currently it can't be a label mismatch but in the future it would be once `performMemberLookup` would stop filtering candidates based on labels. Are you proposing I drop `overloads`, use simple name and for special names just descriptive kind? 

Something like this:

```
error: no exact matches in call to initializer
      self.init(bytes)
           ^
Swift.BinaryInteger:4:5: note: candidate requires that 'S' conform to 'BinaryFloatingPoint' (requirement specified as 'T' == 'BinaryFloatingPoint')
    init<T>(_ source: T) where T : BinaryFloatingPoint
    ^
Swift.BinaryInteger:5:5: note: candidate requires that 'S' conform to 'BinaryInteger' (requirement specified as 'T' == 'BinaryFloatingPoint')
    init<T>(_ source: T) where T : BinaryInteger
    ^
Swift.BinaryInteger:10:10 note: incorrect labels for candidate (expected: '(x:)', have: '(_:)')
    init(bytes: [UInt8])
    ^
```

gregomni commented:
Yes, I think “no exact matches in call” is better, because e.g. in the `extP4a()` test, there aren’t multiple overloads of `extP4a`, so mentioning overloads in the error attached to that line just causes confusion.

xedin commented:
Alright, updated my changes. Thanks!

xedin commented:
@swift-ci please smoke test

jrose-apple commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -2490,8 +2490,16 @@ bool ConstraintSystem::diagnoseAmbiguityWithFixes(
     } else {
-      TC.diagnose(commonAnchor->getLoc(), diag::ambiguous_reference_to_decl,
-                  decl->getDescriptiveKind(), decl->getFullName());
+      auto name = decl->getBaseName();
```
I'd love if this could be the full name in cases where there was only one full name, but I guess that's more work that it's worth right now. (Nice touch with the check for special names, though.)

xedin replied:
Actually it should be that bad, so I'm just going to do it in this PR.

xedin replied:
Done!

xedin committed:
[CSDiagnostics] Split `no exact match` diagnostic into three choices

1. If this is a special name avoid printing it because
   printing kind is sufficient;

2. If all of the labels match, print a full name;

3. If labels in different choices are different, it means
   that we can only print a base name.

xedin commented:
@swift-ci please smoke test
